### ISSUE TITLE
Simulate fingerprint ID on simulators?

### ISSUE THREAD

**Issue by Jonahss** [facebook/WebDriverAgent#26]
According to this, looks like there exists a framework for simulating touch ID: http://stackoverflow.com/questions/25459059/is-there-any-way-to-use-apples-touch-id-fingerprint-scanner-on-ios-simulator

Think we could find the way to call it ourselves?

**Comment by mmmulani** [comment]
it looks like that posts a notification of name "com.apple.BiometricKit_Sim.fingerTouch.match"/"com.apple.BiometricKit_Sim.fingerTouch.nomatch". If that's enough, it should be pretty easy but haven't tested it out. Do you want to try it?

**Comment by Jonahss** [comment]
I've got a couple higher-priority things first I wanna clear out. Working on getting a working webdriveragent-driver for appium. Then I'm going to start looking for unimplemented api endpoints.

**Comment by mmmulani** [comment]
sounds good :)

-Mehdi

**Comment by Jonahss** [comment]
Could you explain how you detected that notification?

**Comment by mmmulani** [comment]
I just opened the simulator in Hopper and started poking around for some
touch ID stuff

**Comment by kaleposhobios** [comment]
I tested this and it seems to work. I added a route like this:
    [[FBRoute POST:@"/touch_id_fail"] respond: ^ id<FBResponsePayload> (FBRouteRequest *request) {
        if (notify_post("com.apple.BiometricKit_Sim.fingerTouch.nomatch")) {
            return FBResponseDictionaryWithOK();
        } else {
            return FBResponseDictionaryWithStatus(FBCommandStatusUnsupported, nil);
        }
    }],
I tested it on my touch id enabled app and it works great. 

I'm not sure which route file you want to add this to, but I confirm the notification you mentioned will work and we can easily implement it.

**Comment by mmmulani** [comment]
I would add a new FBTouchIDCommands.m file here: https://github.com/facebook/WebDriverAgent/tree/master/WebDriverAgentLib/WebDriverCommands and add them

**Comment by Jonahss** [comment]
In appium we have custom routes like this:
`'/wd/hub/session/:sessionId/appium/device/lock'`

You could add a route like
`/session/:sessionId/xcuitest/simulator/touch_id`. Pass a boolean parameter for pass/fail? Or something along those lines.

**Comment by kaleposhobios** [comment]
Why WebDriverAgentLib and not XCTWebDriverAgentLib? Or to put it another way, do I have to duplicate the code? Can I simply add the file to both targets?

**Comment by kaleposhobios** [comment]
I added it as suggested and it seems to work inside XCUITests as-is. I guess I answered my own question.

I tried a number of things to simulate the "enroll finger" menu option but wasn't able to, e.g. posting this Darwin notification: com.apple.BiometricKit.enrollmentChanged

In the hopper I found a selector on GuiController that probably triggers this but I have no clue if we have any way to interact the simulator at that level.
